src/pkg/cli/compose.go

```diff
@@ -206,47 +208,50 @@ func convertPort(port types.ServicePortConfig) (*v1.Port, error) {
 	case "grpc": // TODO: not per spec
 		pbPort.Protocol = v1.Protocol_GRPC
 	default:
-		return nil, fmt.Errorf("port protocol not one of [tcp udp http http2 grpc]: %v", port.Protocol)
+		return nil, fmt.Errorf("port protocol not one of [tcp udp http http2 grpc]: %v", port.Protocol), nil
 	}
 
-	logrus := logrus.WithField("target", port.Target)
+	target := fmt.Sprintf(" target=%d", port.Target)
 
+	var warnings []Warning
 	switch port.Mode {
 	case "":
-		logrus.Warn("No port mode was specified; assuming 'host' (add 'mode' to silence)")
+		warnings = append(warnings, Warning("No port mode was specified; assuming 'host' add 'mode' to silence"+target))
 		HadWarnings = true
 		fallthrough
 	case "host":
 		pbPort.Mode = v1.Mode_HOST
 	case "ingress":
 		// This code is unnecessarily complex because compose-go silently converts short syntax to ingress+tcp
 		if port.Published != "" {
-			logrus.Warn("Published ports are not supported in ingress mode; assuming 'host' (add 'mode' to silence)")
+			warnings = append(warnings, Warning("Published ports are not supported in ingress mode; assuming 'host' (add 'mode' to silence)"+target))
 			HadWarnings = true
 			break
 		}
 		pbPort.Mode = v1.Mode_INGRESS
 		if pbPort.Protocol == v1.Protocol_TCP || pbPort.Protocol == v1.Protocol_UDP {
-			logrus.Warn("TCP ingress is not supported; assuming HTTP")
+			warnings = append(warnings, Warning("TCP ingress is not supported; assuming HTTP"+target))
 			HadWarnings = true
 			pbPort.Protocol = v1.Protocol_HTTP
 		}
 	default:
-		return nil, fmt.Errorf("port mode not one of [host ingress]: %v", port.Mode)
+		return nil, fmt.Errorf("port mode not one of [host ingress]: %v", port.Mode), nil
 	}
-	return pbPort, nil
+	return pbPort, nil, nil
 }
 
-func convertPorts(ports []types.ServicePortConfig) ([]*v1.Port, error) {
```

Comment:
create two versions: one only for checking errors `validatePorts`, and one that converts+warnings

Reply:
done, plz see updated pr